iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_sender.inl

```diff
@@ -0,0 +1,171 @@
+            }
+        }
+        else
+        {
+            return cxx::error<ChunkSenderError>(ChunkSenderError::RUNNING_OUT_OF_CHUNKS);
+        }
+    }
+}
+
+template <typename ChunkDistributorType>
+inline void ChunkSender<ChunkDistributorType>::free(mepoo::ChunkHeader* const chunkHeader) noexcept
+{
+    mepoo::SharedChunk chunk(nullptr);
+    if (!getMembers()->m_chunksInUse.remove(chunkHeader, chunk))
+    {
+        errorHandler(Error::kPOPO__CHUNK_SENDER_INVALID_CHUNK_TO_FREE_FROM_USER, nullptr, ErrorLevel::SEVERE);
+    }
+}
+
+template <typename ChunkDistributorType>
+inline void ChunkSender<ChunkDistributorType>::send(mepoo::ChunkHeader* const chunkHeader) noexcept
+{
+    mepoo::SharedChunk chunk(nullptr);
+    // BEGIN of critical section, chunk will be lost if process gets hard terminated in between
+    if (getChunkReadyForSend(chunkHeader, chunk))
+    {
+        this->deliverToAllStoredQueues(chunk);
+        getMembers()->m_lastChunk = chunk;
+    }
+    // END of critical section, chunk will be lost if process gets hard terminated in between
+}
+
+template <typename ChunkDistributorType>
+inline void ChunkSender<ChunkDistributorType>::pushToHistory(mepoo::ChunkHeader* const chunkHeader) noexcept
+{
+    mepoo::SharedChunk chunk(nullptr);
+    // BEGIN of critical section, chunk will be lost if process gets hard terminated in between
+    if (getChunkReadyForSend(chunkHeader, chunk))
+    {
+        this->addToHistoryWithoutDelivery(chunk);
+        getMembers()->m_lastChunk = chunk;
+    }
+    // END of critical section, chunk will be lost if process gets hard terminated in between
+}
+
+template <typename ChunkDistributorType>
+inline cxx::optional<const mepoo::ChunkHeader*> ChunkSender<ChunkDistributorType>::getLastChunk() const noexcept
+{
+    if (getMembers()->m_lastChunk)
+    {
+        return getMembers()->m_lastChunk.getChunkHeader();
+    }
+    else
+    {
+        return cxx::nullopt_t();
+    }
+}
+
+template <typename ChunkDistributorType>
+inline void ChunkSender<ChunkDistributorType>::releaseAllChunks() noexcept
+{
+    getMembers()->m_chunksInUse.cleanup();
+    this->cleanup();
+    getMembers()->m_lastChunk = nullptr;
+}
+
+template <typename ChunkDistributorType>
+inline bool ChunkSender<ChunkDistributorType>::getChunkReadyForSend(mepoo::ChunkHeader* chunkHeader,
+                                                                    mepoo::SharedChunk& chunk) noexcept
+{
+    if (getMembers()->m_chunksInUse.remove(chunkHeader, chunk))
+    {
+        auto& chunkInfo = chunk.getChunkHeader()->m_info;
+        if (!chunkInfo.m_externalSequenceNumber_bl)
+        {
+            chunkInfo.m_sequenceNumber = getMembers()->m_sequenceNumber;
+            getMembers()->m_sequenceNumber++;
+        }
+        else
+        {
+            getMembers()->m_sequenceNumber++; // for Introspection, else nobody updates.

```

Comment:
It is a bit odd to have special introspection data handling in this class. Could we move this to another place? This method only cares about incrementing the `sequenceNumber`. Maybe we want to rename it?

Reply:
If the sequence number is set on user side, we use this one for the ChunkInfo. But for the throughput calculation stuff, the sequence number that belongs to the port data in shared memory is used. Therefore this must be incremented. I copied this section from the old sender port which will be replaced with the new building blocks. So I would say this is no special introspection data handling but more keeping an internal sequence number updated also if the one that is visible in the chunk will be set by the user. The comment could be better here